File: menubar/Sources/Pulse/UsageClient.swift
import Foundation

struct UsageWindow {
    let utilization: Double  // 0-100 (already in percent units)
    let resetsAt: String?
}

struct UsageData {
    let fiveHour: UsageWindow
    let sevenDay: UsageWindow
    let sevenDayOpus: UsageWindow?
    let sevenDaySonnet: UsageWindow?
}

/// One aligned row of the menu usage table: label | percent | reset text.
struct UsageRow {
    let label: String   // "5h", "Weekly", "Weekly Opus", "Weekly Sonnet"
    let pct: Int         // 0-100
    let reset: String    // already-formatted, e.g. "resets in 4h 26m"
}

enum UsageError: Error, LocalizedError {
    case auth
    case http(Int)
    case rateLimited(retryAfter: TimeInterval?)
    case malformed
    var errorDescription: String? {
        switch self {
        case .auth: return "Authentication expired. Log in again."
        case .http(let c): return "usage API error: HTTP \(c)"
        case .rateLimited: return "usage API error: HTTP 429"
        case .malformed: return "Malformed usage response."
        }
    }
}

/// Extract Retry-After (seconds) from the error. Returns nil if not rateLimited.
func retryAfter(from error: Error) -> TimeInterval? {
    if case UsageError.rateLimited(let ra) = error { return ra }
    return nil
}

private func parseWindow(_ any: Any?) -> UsageWindow? {
    guard let d = any as? [String: Any],
          let num = d["utilization"] as? NSNumber else {
        return nil
    }
    return UsageWindow(utilization: num.doubleValue, resetsAt: d["resets_at"] as? String)
}

/// Raw JSON -> UsageData
func parseUsage(_ json: Any) throws -> UsageData {
    guard let obj = json as? [String: Any],
          let five = parseWindow(obj["five_hour"]),
          let week = parseWindow(obj["seven_day"]) else {
        throw UsageError.malformed
    }
    return UsageData(
        fiveHour: five,
        sevenDay: week,
        sevenDayOpus: parseWindow(obj["seven_day_opus"]),
        sevenDaySonnet: parseWindow(obj["seven_day_sonnet"])
    )
}

func fetchUsage(token: String) async throws -> UsageData {
    var req = URLRequest(url: URL(string: "https://api.anthropic.com/api/oauth/usage")!)
    req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
    req.setValue("oauth-2025-04-20", forHTTPHeaderField: "anthropic-beta")
    req.timeoutInterval = 20

    let (data, resp) = try await URLSession.shared.data(for: req)
    guard let http = resp as? HTTPURLResponse else {
        throw UsageError.malformed
    }
    if http.statusCode == 401 || http.statusCode == 403 {
        throw UsageError.auth
    }
    if http.statusCode == 429 {
        let ra = http.value(forHTTPHeaderField: "Retry-After").flatMap { Int($0) }
        throw UsageError.rateLimited(retryAfter: ra.map { TimeInterval($0) })
    }
    guard (200..<300).contains(http.statusCode) else {
        throw UsageError.http(http.statusCode)
    }
    let json = try JSONSerialization.jsonObject(with: data)
    return try parseUsage(json)
}

/// Fetch usage via the credential cache. On 401/403 the cached token is
/// invalidated and, if the credential store already holds a different token
/// (Claude Code rotated it), the request is retried exactly once.
/// Strictly read-only: never refreshes or writes tokens.
func fetchUsageCached(cache: CredentialCache, force: Bool = false) async throws -> UsageData {
    let token = try cache.getToken(force: force)
    do {
        return try await fetchUsage(token: token)
    } catch UsageError.auth {
        cache.invalidate()
        guard let retryToken = try? cache.getToken(), retryToken != token else {
            throw UsageError.auth
        }
        do {
            return try await fetchUsage(token: retryToken)
        } catch UsageError.auth {
            cache.invalidate()
            throw UsageError.auth
        }
    }
}

File: src/usageClient.ts
import { defaultCache, type CredentialCache } from "./credentialCache";

const USAGE_URL = "https://api.anthropic.com/api/oauth/usage";

export interface UsageWindow {
  /** percent, 0-100 */
  utilization: number;
  /** ISO 8601, may be absent */
  resetsAt: string | null;
}

export interface UsageData {
  fiveHour: UsageWindow;
  sevenDay: UsageWindow;
  sevenDayOpus: UsageWindow | null;
  sevenDaySonnet: UsageWindow | null;
}

export class AuthError extends Error {}

/** Retryable transient error (network/5xx/429, etc.). */
export class TransientError extends Error {
  readonly retryAfterMs?: number;
  constructor(message: string, retryAfterMs?: number) {
    super(message);
    this.retryAfterMs = retryAfterMs;
  }
}

/** Convert the Retry-After header (integer seconds) to ms. undefined if absent or non-integer (e.g. HTTP-date). */
export function parseRetryAfterMs(header: string | null): number | undefined {
  if (header == null) return undefined;
  const trimmed = header.trim();
  if (!/^\d+$/.test(trimmed)) return undefined;
  return parseInt(trimmed, 10) * 1000;
}

function parseWindow(raw: unknown): UsageWindow | null {
  if (!raw || typeof raw !== "object") {
    return null;
  }
  const obj = raw as Record<string, unknown>;
  const util = obj.utilization;
  if (typeof util !== "number") {
    return null;
  }
  const resets = obj.resets_at;
  return {
    utilization: util,
    resetsAt: typeof resets === "string" ? resets : null,
  };
}

/** Convert raw API JSON into UsageData (pure function, tested). */
export function parseUsage(json: unknown): UsageData {
  const obj = (json && typeof json === "object" ? json : {}) as Record<string, unknown>;
  const fiveHour = parseWindow(obj.five_hour);
  const sevenDay = parseWindow(obj.seven_day);
  if (!fiveHour || !sevenDay) {
    throw new Error("usage response is missing five_hour/seven_day.");
  }
  return {
    fiveHour,
    sevenDay,
    sevenDayOpus: parseWindow(obj.seven_day_opus),
    sevenDaySonnet: parseWindow(obj.seven_day_sonnet),
  };
}

async function requestUsage(token: string): Promise<{ status: number; data?: UsageData }> {
  const res = await fetch(USAGE_URL, {
    headers: {
      Authorization: `Bearer ${token}`,
      "anthropic-beta": "oauth-2025-04-20",
    },
  });

  if (res.status === 401 || res.status === 403) {
    return { status: res.status };
  }
  if (res.status === 429) {
    throw new TransientError(
      `usage API error: HTTP 429`,
      parseRetryAfterMs(res.headers.get("retry-after"))
    );
  }
  if (!res.ok) {
    throw new TransientError(`usage API error: HTTP ${res.status}`);
  }

  return { status: res.status, data: parseUsage(await res.json()) };
}

/**
 * Call the usage endpoint to fetch current usage.
 * On 401/403 the cached token is invalidated and, if the credential store
 * already holds a different token (Claude Code rotated it), the request is
 * retried exactly once. Strictly read-only: never refreshes or writes tokens.
 */
export async function fetchUsage(force = false, cache: CredentialCache = defaultCache): Promise<UsageData> {
  const token = await cache.getToken(force);
  const first = await requestUsage(token);
  if (first.data) {
    return first.data;
  }

  cache.invalidate();
  let retryToken: string;
  try {
    retryToken = await cache.getToken();
  } catch {
    throw new AuthError("Authentication expired. Log in again.");
  }
  if (retryToken === token) {
    throw new AuthError("Authentication expired. Log in again.");
  }
  const second = await requestUsage(retryToken);
  if (second.data) {
    return second.data;
  }
  cache.invalidate();
  throw new AuthError("Authentication expired. Log in again.");
}
